// @chakra-ui/stat/dist/index.cjs.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Stat: () => Stat,
  StatArrow: () => StatArrow,
  StatDownArrow: () => StatDownArrow,
  StatGroup: () => StatGroup,
  StatHelpText: () => StatHelpText,
  StatLabel: () => StatLabel,
  StatNumber: () => StatNumber,
  StatUpArrow: () => StatUpArrow,
  useStatStyles: () => useStatStyles
});
module.exports = __toCommonJS(src_exports);

// ../../react-shim.js
var import_react = __toESM(require("react"));

// src/stat.tsx
var import_icon = require("@chakra-ui/icon");
var import_react_utils = require("@chakra-ui/react-utils");
var import_system = require("@chakra-ui/system");
var import_utils = require("@chakra-ui/utils");
var import_visually_hidden = require("@chakra-ui/visually-hidden");
var [StatStylesProvider, useStatStyles] = (0, import_react_utils.createContext)({
  name: `StatStylesContext`,
  errorMessage: `useStatStyles returned is 'undefined'. Seems you forgot to wrap the components in "<Stat />" `
});
var StatLabel = (0, import_system.forwardRef)(function StatLabel2(props, ref) {
  const styles = useStatStyles();
  return /* @__PURE__ */ import_react.default.createElement(import_system.chakra.dt, {
    ref,
    ...props,
    className: (0, import_utils.cx)("chakra-stat__label", props.className),
    __css: styles.label
  });
});
if (import_utils.__DEV__) {
  StatLabel.displayName = "StatLabel";
}
var StatHelpText = (0, import_system.forwardRef)(function StatHelpText2(props, ref) {
  const styles = useStatStyles();
  return /* @__PURE__ */ import_react.default.createElement(import_system.chakra.dd, {
    ref,
    ...props,
    className: (0, import_utils.cx)("chakra-stat__help-text", props.className),
    __css: styles.helpText
  });
});
if (import_utils.__DEV__) {
  StatHelpText.displayName = "StatHelpText";
}
var StatNumber = (0, import_system.forwardRef)(function StatNumber2(props, ref) {
  const styles = useStatStyles();
  return /* @__PURE__ */ import_react.default.createElement(import_system.chakra.dd, {
    ref,
    ...props,
    className: (0, import_utils.cx)("chakra-stat__number", props.className),
    __css: {
      ...styles.number,
      fontFeatureSettings: "pnum",
      fontVariantNumeric: "proportional-nums"
    }
  });
});
if (import_utils.__DEV__) {
  StatNumber.displayName = "StatNumber";
}
var StatDownArrow = (props) => /* @__PURE__ */ import_react.default.createElement(import_icon.Icon, {
  color: "red.400",
  ...props
}, /* @__PURE__ */ import_react.default.createElement("path", {
  fill: "currentColor",
  d: "M21,5H3C2.621,5,2.275,5.214,2.105,5.553C1.937,5.892,1.973,6.297,2.2,6.6l9,12 c0.188,0.252,0.485,0.4,0.8,0.4s0.611-0.148,0.8-0.4l9-12c0.228-0.303,0.264-0.708,0.095-1.047C21.725,5.214,21.379,5,21,5z"
}));
if (import_utils.__DEV__) {
  StatDownArrow.displayName = "StatDownArrow";
}
function StatUpArrow(props) {
  return /* @__PURE__ */ import_react.default.createElement(import_icon.Icon, {
    color: "green.400",
    ...props
  }, /* @__PURE__ */ import_react.default.createElement("path", {
    fill: "currentColor",
    d: "M12.8,5.4c-0.377-0.504-1.223-0.504-1.6,0l-9,12c-0.228,0.303-0.264,0.708-0.095,1.047 C2.275,18.786,2.621,19,3,19h18c0.379,0,0.725-0.214,0.895-0.553c0.169-0.339,0.133-0.744-0.095-1.047L12.8,5.4z"
  }));
}
if (import_utils.__DEV__) {
  StatUpArrow.displayName = "StatUpArrow";
}
function StatArrow(props) {
  const { type, "aria-label": ariaLabel, ...rest } = props;
  const styles = useStatStyles();
  const BaseIcon = type === "increase" ? StatUpArrow : StatDownArrow;
  const defaultAriaLabel = type === "increase" ? "increased by" : "decreased by";
  const label = ariaLabel || defaultAriaLabel;
  return /* @__PURE__ */ import_react.default.createElement(import_react.default.Fragment, null, /* @__PURE__ */ import_react.default.createElement(import_visually_hidden.VisuallyHidden, null, label), /* @__PURE__ */ import_react.default.createElement(BaseIcon, {
    "aria-hidden": true,
    ...rest,
    __css: styles.icon
  }));
}
if (import_utils.__DEV__) {
  StatArrow.displayName = "StatArrow";
}
var Stat = (0, import_system.forwardRef)(function Stat2(props, ref) {
  const styles = (0, import_system.useMultiStyleConfig)("Stat", props);
  const statStyles = {
    position: "relative",
    flex: "1 1 0%",
    ...styles.container
  };
  const { className, children, ...rest } = (0, import_system.omitThemingProps)(props);
  return /* @__PURE__ */ import_react.default.createElement(StatStylesProvider, {
    value: styles
  }, /* @__PURE__ */ import_react.default.createElement(import_system.chakra.div, {
    ref,
    ...rest,
    className: (0, import_utils.cx)("chakra-stat", className),
    __css: statStyles
  }, /* @__PURE__ */ import_react.default.createElement("dl", null, children)));
});
if (import_utils.__DEV__) {
  Stat.displayName = "Stat";
}
var StatGroup = (0, import_system.forwardRef)(function StatGroup2(props, ref) {
  return /* @__PURE__ */ import_react.default.createElement(import_system.chakra.div, {
    ...props,
    ref,
    role: "group",
    className: (0, import_utils.cx)("chakra-stat__group", props.className),
    __css: {
      display: "flex",
      flexWrap: "wrap",
      justifyContent: "space-around",
      alignItems: "flex-start"
    }
  });
});
if (import_utils.__DEV__) {
  StatGroup.displayName = "StatGroup";
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Stat,
  StatArrow,
  StatDownArrow,
  StatGroup,
  StatHelpText,
  StatLabel,
  StatNumber,
  StatUpArrow,
  useStatStyles
});
