// @chakra-ui/radio/dist/index.cjs.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Radio: () => Radio,
  RadioGroup: () => RadioGroup,
  useRadio: () => useRadio,
  useRadioGroup: () => useRadioGroup,
  useRadioGroupContext: () => useRadioGroupContext
});
module.exports = __toCommonJS(src_exports);

// ../../react-shim.js
var import_react = __toESM(require("react"));

// src/radio.tsx
var import_system2 = require("@chakra-ui/system");
var import_utils4 = require("@chakra-ui/utils");

// src/radio-group.tsx
var import_system = require("@chakra-ui/system");
var import_utils2 = require("@chakra-ui/utils");
var import_react_utils2 = require("@chakra-ui/react-utils");

// src/use-radio-group.ts
var import_hooks = require("@chakra-ui/hooks");
var import_utils = require("@chakra-ui/utils");
var import_react_utils = require("@chakra-ui/react-utils");
var import_react2 = require("react");
function useRadioGroup(props = {}) {
  const {
    onChange: onChangeProp,
    value: valueProp,
    defaultValue,
    name: nameProp,
    isDisabled,
    isFocusable,
    isNative,
    ...htmlProps
  } = props;
  const [valueState, setValue] = (0, import_react2.useState)(defaultValue || "");
  const [isControlled, value] = (0, import_hooks.useControllableProp)(valueProp, valueState);
  const ref = (0, import_react2.useRef)(null);
  const focus = (0, import_react2.useCallback)(() => {
    const rootNode = ref.current;
    if (!rootNode)
      return;
    let query = `input:not(:disabled):checked`;
    const firstEnabledAndCheckedInput = rootNode.querySelector(query);
    if (firstEnabledAndCheckedInput) {
      firstEnabledAndCheckedInput.focus();
      return;
    }
    query = `input:not(:disabled)`;
    const firstEnabledInput = rootNode.querySelector(query);
    firstEnabledInput == null ? void 0 : firstEnabledInput.focus();
  }, []);
  const fallbackName = (0, import_hooks.useId)(void 0, `radio`);
  const name = nameProp || fallbackName;
  const onChange = (0, import_react2.useCallback)((eventOrValue) => {
    const nextValue = (0, import_utils.isInputEvent)(eventOrValue) ? eventOrValue.target.value : eventOrValue;
    if (!isControlled) {
      setValue(nextValue);
    }
    onChangeProp == null ? void 0 : onChangeProp(String(nextValue));
  }, [onChangeProp, isControlled]);
  const getRootProps = (0, import_react2.useCallback)((props2 = {}, forwardedRef = null) => ({
    ...props2,
    ref: (0, import_react_utils.mergeRefs)(forwardedRef, ref),
    role: "radiogroup"
  }), []);
  const getRadioProps = (0, import_react2.useCallback)((props2 = {}, ref2 = null) => {
    const checkedKey = isNative ? "checked" : "isChecked";
    return {
      ...props2,
      ref: ref2,
      name,
      [checkedKey]: value != null ? props2.value === value : void 0,
      onChange,
      "data-radiogroup": true
    };
  }, [isNative, name, onChange, value]);
  return {
    getRootProps,
    getRadioProps,
    name,
    ref,
    focus,
    setValue,
    value,
    onChange,
    isDisabled,
    isFocusable,
    htmlProps
  };
}

// src/radio-group.tsx
var import_react3 = require("react");
var [RadioGroupProvider, useRadioGroupContext] = (0, import_react_utils2.createContext)({
  name: "RadioGroupContext",
  strict: false
});
var RadioGroup = (0, import_system.forwardRef)((props, ref) => {
  const {
    colorScheme,
    size,
    variant,
    children,
    className,
    isDisabled,
    isFocusable,
    ...rest
  } = props;
  const { value, onChange, getRootProps, name, htmlProps } = useRadioGroup(rest);
  const group = (0, import_react3.useMemo)(() => ({
    name,
    size,
    onChange,
    colorScheme,
    value,
    variant,
    isDisabled,
    isFocusable
  }), [
    name,
    size,
    onChange,
    colorScheme,
    value,
    variant,
    isDisabled,
    isFocusable
  ]);
  const groupProps = getRootProps(htmlProps, ref);
  const _className = (0, import_utils2.cx)("chakra-radio-group", className);
  return /* @__PURE__ */ import_react.default.createElement(RadioGroupProvider, {
    value: group
  }, /* @__PURE__ */ import_react.default.createElement(import_system.chakra.div, {
    ...groupProps,
    className: _className
  }, children));
});
if (import_utils2.__DEV__) {
  RadioGroup.displayName = "RadioGroup";
}

// src/use-radio.ts
var import_form_control = require("@chakra-ui/form-control");
var import_hooks2 = require("@chakra-ui/hooks");
var import_utils3 = require("@chakra-ui/utils");
var import_visually_hidden = require("@chakra-ui/visually-hidden");
var import_focus_visible = require("@zag-js/focus-visible");
var import_react4 = require("react");
function useRadio(props = {}) {
  const {
    defaultChecked,
    isChecked: isCheckedProp,
    isFocusable,
    isDisabled: isDisabledProp,
    isReadOnly: isReadOnlyProp,
    isRequired: isRequiredProp,
    onChange,
    isInvalid: isInvalidProp,
    name,
    value,
    id: idProp,
    "data-radiogroup": dataRadioGroup,
    "aria-describedby": ariaDescribedBy,
    ...htmlProps
  } = props;
  const uuid = (0, import_hooks2.useId)(void 0, "radio");
  const formControl = (0, import_form_control.useFormControlContext)();
  const group = useRadioGroupContext();
  const isWithinRadioGroup = !!group || !!dataRadioGroup;
  const isWithinFormControl = !!formControl;
  let id = isWithinFormControl && !isWithinRadioGroup ? formControl.id : uuid;
  id = idProp ?? id;
  const isDisabled = isDisabledProp ?? (formControl == null ? void 0 : formControl.isDisabled);
  const isReadOnly = isReadOnlyProp ?? (formControl == null ? void 0 : formControl.isReadOnly);
  const isRequired = isRequiredProp ?? (formControl == null ? void 0 : formControl.isRequired);
  const isInvalid = isInvalidProp ?? (formControl == null ? void 0 : formControl.isInvalid);
  const [isFocusVisible, setIsFocusVisible] = (0, import_react4.useState)(false);
  const [isFocused, setFocused] = (0, import_hooks2.useBoolean)();
  const [isHovered, setHovering] = (0, import_hooks2.useBoolean)();
  const [isActive, setActive] = (0, import_hooks2.useBoolean)();
  const [isCheckedState, setChecked] = (0, import_react4.useState)(Boolean(defaultChecked));
  const [isControlled, isChecked] = (0, import_hooks2.useControllableProp)(isCheckedProp, isCheckedState);
  (0, import_react4.useEffect)(() => {
    return (0, import_focus_visible.trackFocusVisible)(setIsFocusVisible);
  }, []);
  const handleChange = (0, import_react4.useCallback)((event) => {
    if (isReadOnly || isDisabled) {
      event.preventDefault();
      return;
    }
    if (!isControlled) {
      setChecked(event.target.checked);
    }
    onChange == null ? void 0 : onChange(event);
  }, [isControlled, isDisabled, isReadOnly, onChange]);
  const onKeyDown = (0, import_react4.useCallback)((event) => {
    if (event.key === " ") {
      setActive.on();
    }
  }, [setActive]);
  const onKeyUp = (0, import_react4.useCallback)((event) => {
    if (event.key === " ") {
      setActive.off();
    }
  }, [setActive]);
  const getRadioProps = (0, import_react4.useCallback)((props2 = {}, ref = null) => ({
    ...props2,
    ref,
    "data-active": (0, import_utils3.dataAttr)(isActive),
    "data-hover": (0, import_utils3.dataAttr)(isHovered),
    "data-disabled": (0, import_utils3.dataAttr)(isDisabled),
    "data-invalid": (0, import_utils3.dataAttr)(isInvalid),
    "data-checked": (0, import_utils3.dataAttr)(isChecked),
    "data-focus": (0, import_utils3.dataAttr)(isFocused),
    "data-focus-visible": (0, import_utils3.dataAttr)(isFocused && isFocusVisible),
    "data-readonly": (0, import_utils3.dataAttr)(isReadOnly),
    "aria-hidden": true,
    onMouseDown: (0, import_utils3.callAllHandlers)(props2.onMouseDown, setActive.on),
    onMouseUp: (0, import_utils3.callAllHandlers)(props2.onMouseUp, setActive.off),
    onMouseEnter: (0, import_utils3.callAllHandlers)(props2.onMouseEnter, setHovering.on),
    onMouseLeave: (0, import_utils3.callAllHandlers)(props2.onMouseLeave, setHovering.off)
  }), [
    isActive,
    isHovered,
    isDisabled,
    isInvalid,
    isChecked,
    isFocused,
    isReadOnly,
    setActive.on,
    setActive.off,
    setHovering.on,
    setHovering.off,
    isFocusVisible
  ]);
  const { onFocus, onBlur } = formControl ?? {};
  const getInputProps = (0, import_react4.useCallback)((props2 = {}, ref = null) => {
    const trulyDisabled = isDisabled && !isFocusable;
    return {
      ...props2,
      id,
      ref,
      type: "radio",
      name,
      value,
      onChange: (0, import_utils3.callAllHandlers)(props2.onChange, handleChange),
      onBlur: (0, import_utils3.callAllHandlers)(onBlur, props2.onBlur, setFocused.off),
      onFocus: (0, import_utils3.callAllHandlers)(onFocus, props2.onFocus, setFocused.on),
      onKeyDown: (0, import_utils3.callAllHandlers)(props2.onKeyDown, onKeyDown),
      onKeyUp: (0, import_utils3.callAllHandlers)(props2.onKeyUp, onKeyUp),
      checked: isChecked,
      disabled: trulyDisabled,
      readOnly: isReadOnly,
      required: isRequired,
      "aria-invalid": (0, import_utils3.ariaAttr)(isInvalid),
      "aria-disabled": (0, import_utils3.ariaAttr)(trulyDisabled),
      "aria-required": (0, import_utils3.ariaAttr)(isRequired),
      "data-readonly": (0, import_utils3.dataAttr)(isReadOnly),
      "aria-describedby": ariaDescribedBy,
      style: import_visually_hidden.visuallyHiddenStyle
    };
  }, [
    isDisabled,
    isFocusable,
    id,
    name,
    value,
    handleChange,
    onBlur,
    setFocused,
    onFocus,
    onKeyDown,
    onKeyUp,
    isChecked,
    isReadOnly,
    isRequired,
    isInvalid,
    ariaDescribedBy
  ]);
  const getLabelProps = (props2 = {}, ref = null) => ({
    ...props2,
    ref,
    onMouseDown: (0, import_utils3.callAllHandlers)(props2.onMouseDown, stop),
    onTouchStart: (0, import_utils3.callAllHandlers)(props2.onTouchStart, stop),
    "data-disabled": (0, import_utils3.dataAttr)(isDisabled),
    "data-checked": (0, import_utils3.dataAttr)(isChecked),
    "data-invalid": (0, import_utils3.dataAttr)(isInvalid)
  });
  const getRootProps = (props2, ref = null) => ({
    ...props2,
    ref,
    "data-disabled": (0, import_utils3.dataAttr)(isDisabled),
    "data-checked": (0, import_utils3.dataAttr)(isChecked),
    "data-invalid": (0, import_utils3.dataAttr)(isInvalid)
  });
  const state = {
    isInvalid,
    isFocused,
    isChecked,
    isActive,
    isHovered,
    isDisabled,
    isReadOnly,
    isRequired
  };
  return {
    state,
    getCheckboxProps: getRadioProps,
    getInputProps,
    getLabelProps,
    getRootProps,
    htmlProps
  };
}
function stop(event) {
  event.preventDefault();
  event.stopPropagation();
}

// src/radio.tsx
var Radio = (0, import_system2.forwardRef)((props, ref) => {
  const group = useRadioGroupContext();
  const { onChange: onChangeProp, value: valueProp } = props;
  const styles = (0, import_system2.useMultiStyleConfig)("Radio", { ...group, ...props });
  const ownProps = (0, import_system2.omitThemingProps)(props);
  const {
    spacing = "0.5rem",
    children,
    isDisabled = group == null ? void 0 : group.isDisabled,
    isFocusable = group == null ? void 0 : group.isFocusable,
    inputProps: htmlInputProps,
    ...rest
  } = ownProps;
  let isChecked = props.isChecked;
  if ((group == null ? void 0 : group.value) != null && valueProp != null) {
    isChecked = group.value === valueProp;
  }
  let onChange = onChangeProp;
  if ((group == null ? void 0 : group.onChange) && valueProp != null) {
    onChange = (0, import_utils4.callAll)(group.onChange, onChangeProp);
  }
  const name = (props == null ? void 0 : props.name) ?? (group == null ? void 0 : group.name);
  const {
    getInputProps,
    getCheckboxProps,
    getLabelProps,
    getRootProps,
    htmlProps
  } = useRadio({
    ...rest,
    isChecked,
    isFocusable,
    isDisabled,
    onChange,
    name
  });
  const [layoutProps, otherProps] = (0, import_utils4.split)(htmlProps, import_system2.layoutPropNames);
  const checkboxProps = getCheckboxProps(otherProps);
  const inputProps = getInputProps(htmlInputProps, ref);
  const labelProps = getLabelProps();
  const rootProps = Object.assign({}, layoutProps, getRootProps());
  const rootStyles = {
    display: "inline-flex",
    alignItems: "center",
    verticalAlign: "top",
    cursor: "pointer",
    position: "relative",
    ...styles.container
  };
  const checkboxStyles = {
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    flexShrink: 0,
    ...styles.control
  };
  const labelStyles = {
    userSelect: "none",
    marginStart: spacing,
    ...styles.label
  };
  return /* @__PURE__ */ import_react.default.createElement(import_system2.chakra.label, {
    className: "chakra-radio",
    ...rootProps,
    __css: rootStyles
  }, /* @__PURE__ */ import_react.default.createElement("input", {
    className: "chakra-radio__input",
    ...inputProps
  }), /* @__PURE__ */ import_react.default.createElement(import_system2.chakra.span, {
    className: "chakra-radio__control",
    ...checkboxProps,
    __css: checkboxStyles
  }), children && /* @__PURE__ */ import_react.default.createElement(import_system2.chakra.span, {
    className: "chakra-radio__label",
    ...labelProps,
    __css: labelStyles
  }, children));
});
if (import_utils4.__DEV__) {
  Radio.displayName = "Radio";
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Radio,
  RadioGroup,
  useRadio,
  useRadioGroup,
  useRadioGroupContext
});
